### ISSUE TITLE
tf code question?

### ISSUE THREAD

**Issue by science-dojo** [kimiyoung/transformer-xl#13]
Good works! I have two question about your tf codes.

The first:
In the paper, query vector is calculated using previous layer's hidden state rather than the concatenated pre-layer's memory and hidden state. However, in the tf code, I found the query vector is calculated the same as key vector and value vector. 

![image](https://user-images.githubusercontent.com/13633999/51364026-86b85000-1b15-11e9-93f7-390051ca7285.png)
![image](https://user-images.githubusercontent.com/13633999/51364180-368dbd80-1b16-11e9-9ce5-8a78845d92ea.png)

The second:
Each layers has memory tensor with shape [mem_len, batch_size, d_model].  When calculating query, key and value vector, the input vector of `tf.layers.dens` is the concatenation of current layer's memory and pre-layers' output. which seems be conflict with the paper. Besides, why stop gradient in `_cache_mem` method rather than in `rel_multihead_attn`, the later seems to make better sense.

**Comment by zihangdai** [comment]
Check out the next line `w_head_q = w_head_q[-qlen:]`, which only take the last part to form the query vector. The initial computation is intended to put everything into a single matrix multiplication, which could be faster on GPU/TPU. 

Stop gradient in `_cache_mem` makes sure there is no more dependency on the segment in the computational graph. This will allow proper garbage collection.

**Comment by guotong1988** [comment]
mark
